src/Algebra/Graph/LabelledAdjacencyMap.hs

```diff
@@ -0,0 +1,780 @@
+-- @
+edgeCount :: LabelledAdjacencyMap a e -> Int
+edgeCount = Map.foldr (\es r -> (Map.size es + r)) 0 . labelledAdjacencyMap
+
+-- | The sorted list of vertices of a given graph.
+-- Complexity: /O(n)/ time and memory.
+--
+-- @
+-- vertexList 'empty'      == []
+-- vertexList ('vertex' x) == [x]
+-- vertexList . 'vertices' == 'Data.List.nub' . 'Data.List.sort'
+-- @
+vertexList :: LabelledAdjacencyMap a e -> [a]
+vertexList = Map.keys . labelledAdjacencyMap
+
+-- | The sorted list of edges of a graph.
+-- Complexity: /O(n + m)/ time and /O(m)/ memory.
+--
+-- @
+-- edgeList 'empty'          == []
+-- edgeList ('vertex' x)     == []
+-- edgeList ('edge' x y)     == [(x,y)]
+-- edgeList ('star' 2 [3,1]) == [(2,1), (2,3)]
+-- edgeList . 'edges'        == 'Data.List.nub' . 'Data.List.sort'
+-- edgeList . 'transpose'    == 'Data.List.sort' . map 'Data.Tuple.swap' . edgeList
+-- @
+edgeList :: LabelledAdjacencyMap a e -> [(a, a)]
+edgeList (LAM m) = do
+        (x, ys) <- Map.toAscList m
+        (y, _ ) <- Map.toAscList ys
+        pure (x, y)
+
+-- | The sorted /adjacency list/ of a graph.
+-- Complexity: /O(n + m)/ time and /O(m)/ memory.
+--
+-- @
+-- adjacencyList 'empty'               == []
+-- adjacencyList ('vertex' x)          == [(x, [])]
+-- adjacencyList ('edge' 1 2)          == [(1, [2]), (2, [])]
+-- adjacencyList ('star' 2 [3,1])      == [(1, []), (2, [1,3]), (3, [])]
+-- 'fromAdjacencyList' . adjacencyList == id
+-- @
+adjacencyList :: LabelledAdjacencyMap a e -> [(a, [a])]
+adjacencyList =
+        map (fmap (map fst . Map.toAscList))
+                . Map.toAscList
+                . labelledAdjacencyMap
+
+-- | The /preset/ of an element @x@ is the set of its /direct predecessors/.
+-- Complexity: /O(n * log(n))/ time and /O(n)/ memory.
+--
+-- @
+-- preSet x 'empty'      == Set.'Set.empty'
+-- preSet x ('vertex' x) == Set.'Set.empty'
+-- preSet 1 ('edge' 1 2) == Set.'Set.empty'
+-- preSet y ('edge' x y) == Set.'Set.fromList' [x]
+-- @
+preSet :: Ord a => a -> LabelledAdjacencyMap a e -> Set.Set a
+preSet x = Set.fromAscList . map fst . filter p . Map.toAscList . labelledAdjacencyMap
+  where
+    p (_, set) = x `Map.member` set
+
+
+
+
+-- | The set of vertices of a given graph.
+-- Complexity: /O(n)/ time and memory.
+--
+-- @
+-- vertexSet 'empty'      == Set.'Set.empty'
+-- vertexSet . 'vertex'   == Set.'Set.singleton'
+-- vertexSet . 'vertices' == Set.'Set.fromList'
+-- vertexSet . 'clique'   == Set.'Set.fromList'
+-- @
+vertexSet :: LabelledAdjacencyMap a e -> Set a
+vertexSet = Map.keysSet . labelledAdjacencyMap
+
+-- | The set of edges of a given graph.
+-- Complexity: /O((n + m) * log(m))/ time and /O(m)/ memory.
+--
+-- @
+-- edgeSet 'empty'      == Set.'Set.empty'
+-- edgeSet ('vertex' x) == Set.'Set.empty'
+-- edgeSet ('edge' x y) == Set.'Set.singleton' (x,y)
+-- edgeSet . 'edges'    == Set.'Set.fromList'
+-- @
+edgeSet :: Ord a => LabelledAdjacencyMap a e -> Set (a, a)
+edgeSet =
```

Comment:
This could be simply `edgeSet = Set.fromAscList . edgeList`, as in `AdjacencyMap`. Or is this implementation faster?

Reply:
I doubt it's faster, it was based on one of the previous implementations in the `AdjacncyMap` implementation.